frontend/catalyst/preprocess.py

```diff
@@ -16,16 +16,116 @@
 import jax
 import pennylane as qml
 from pennylane import transform
-from pennylane.measurements import CountsMP, ExpectationMP, ProbabilityMP, VarianceMP
+from pennylane.measurements import (
+    CountsMP,
+    ExpectationMP,
+    MidMeasureMP,
+    ProbabilityMP,
+    VarianceMP,
+)
 from pennylane.tape.tape import (
     _validate_computational_basis_sampling,
     rotations_and_diagonal_measurements,
 )
 
 import catalyst
+import catalyst.pennylane_extensions
 from catalyst.utils.exceptions import CompileError
 
 
+def _operator_decomposition_gen(
+    op: qml.operation.Operator,
+    acceptance_function,
+    decomposer,
+    max_expansion=None,
+    current_depth=0,
+):
+    """A generator that yields the next operation that is accepted."""
+    max_depth_reached = False
+    if max_expansion is not None and max_expansion <= current_depth:
+        max_depth_reached = True
+    if acceptance_function(op) or max_depth_reached:
+        yield op
+    else:
+        try:
+            decomp = decomposer(op)
+            current_depth += 1
+        except qml.operation.DecompositionUndefinedError as e:
+            raise CompileError(
+                f"Operator {op} not supported on device and does not provide a decomposition."
+            ) from e
+
+        for sub_op in decomp:
+            yield from _operator_decomposition_gen(
+                sub_op,
+                acceptance_function,
+                decomposer=decomposer,
+                max_expansion=max_expansion,
+                current_depth=current_depth,
+            )
+
+
+@transform
+def decompose(
```

Comment:
Very similar to PL decompose

Reply:
Have you considered outlining some parts of this method to make it more readable. For example:

```python
def decompose_to_matrix(op):
   if !op.name in {"MultiControlledX", "BlockEncode"} ...
       return None
   ... # Continue with normal path
   
def decompose_hybrid_op(op):
   if !isinstance(op, HybridOp):
       return None
   ... # Continue with normal path
```

and then

```python
def decomposer(op):
   success = decompose_to_matrix(op)
   if success:
       return success
   success = decompose_hybrid_op(op)
   if success:
       return success
   return op.decomposition()
```

Also, is there a reason why the decomposer needs to be a nested function? Are you making a closure out of some arguments?

Reply:
Ok no worries I do that. Because decomposition potentially need a big depth, template to some high level op to a universal gate set. You check the decompose in PL it is very similar.

Reply:
Changes done!